mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,416 @@
+mod block;
+pub mod error;
+mod mmap;
+
+use std::{
+    collections::VecDeque,
+    fs::{File, OpenOptions},
+    future::Future,
+    path::PathBuf,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::{Context, Poll, Waker},
+    time::Instant,
+};
+
+use crate::persist::storage::{
+    serialize::{binary_deserialize, binary_serialize, binary_serialize_size},
+    FlushOptions, FlushState, Storage, StorageResult,
+};
+use memmap::MmapMut;
+use parking_lot::Mutex;
+
+use std::io::Result as IOResult;
+
+use block::{calculate_hash, validate, BlockHeaderWithHash, Data};
+use error::RingBufferError;
+use mmap::{mmap_read, mmap_write};
+
+pub type RingBufferResult<T> = Result<T, error::RingBufferError>;
+
+#[derive(Default)]
+pub(crate) struct FilePointers {
+    write: AtomicUsize,
+    read: AtomicUsize,
+}
+
+/// Imagine there are three pointers:
+/// W - A write pointer
+/// R - A read pointer
+/// D - A delete pointer
+///
+/// With a file buffer similar to below:
+/// +---+---+---+---+---+---+---+---+---+
+/// | A | B | C | D | E | F | G | H | I |
+/// +---+---+---+---+---+---+---+---+---+
+///
+/// It is possible that W could wrap around
+/// and cause us to lose data due to variable
+/// sized data being written.
+/// If W + data size > R and should write precede read,
+/// the data could become corrupted.
+/// Either this is prevented by scanning the file for
+/// the next block and checking if the write will fit.
+/// Or, W + data must not pass both D and R.
+/// Not passing D helps with guaranteeing all data read was handled.
+/// Not passing R helps avoid corruption.
+pub struct RingBuffer {
+    inner: Arc<Mutex<MmapMut>>,
+    max_file_size: usize,
+    flush_options: FlushOptions,
+    pointers: Arc<FilePointers>,
+    waker: Option<Waker>,
```

Comment:
I am not super confident in my async await ability... but I think now that I have changed the interface on StreamWakeableState we don't need this field or the set_waker fns anymore? Can someone confirm this for me?

Reply:
If RingBuffer is the one who implements `StreamWakeableState`, then you probably still need waker. And I believe it should be used in the following way:
- If a Write future can't do the job (return "Pending" state), you need to clone the waker and remember it in the ring buffer.
- The write future then goes to sleep.
- Next time when the "read" operation is performed, you need to check if there is a waker, and if there is one - wake it.

It seems that you would need two wakers - one for "write" and one for "delete".

Reply:
So, I would need check both wakers in the "read" operation?